src/core/ext/filters/client_channel/resolver/google_c2p/google_c2p_resolver.cc

```diff
@@ -0,0 +1,360 @@
+    RefCountedPtr<GoogleCloud2ProdResolver> resolver,
+    grpc_polling_entity* pollent)
+    : MetadataQuery(std::move(resolver),
+                    "/computeMetadata/v1/instance/network-interfaces/0/ipv6s",
+                    pollent) {}
+
+void GoogleCloud2ProdResolver::IPv6Query::OnDone(
+    GoogleCloud2ProdResolver* resolver, const grpc_http_response* response,
+    grpc_error* error) {
+  if (error != GRPC_ERROR_NONE) {
+    gpr_log(GPR_ERROR, "error fetching IPv6 address from metadata server: %s",
+            grpc_error_string(error));
+  }
+  resolver->IPv6QueryDone(error == GRPC_ERROR_NONE && response->status == 200);
+  GRPC_ERROR_UNREF(error);
+}
+
+//
+// GoogleCloud2ProdResolver
+//
+
+GoogleCloud2ProdResolver::GoogleCloud2ProdResolver(ResolverArgs args)
+    : work_serializer_(std::move(args.work_serializer)),
+      pollent_(grpc_polling_entity_create_from_pollset_set(args.pollset_set)) {
+  absl::string_view name_to_resolve = absl::StripPrefix(args.uri.path(), "/");
+  // If we're not running on GCP, we can't use DirectPath, so delegate
+  // to the DNS resolver.
+  if (!grpc_alts_is_running_on_gcp() ||
+      // If the client is already using xDS, we can't use it here, because
+      // they may be talking to a completely different xDS server than we
+      // want to.
+      // TODO(roth): When we implement xDS federation, remove this constraint.
+      UniquePtr<char>(gpr_getenv("GRPC_XDS_BOOTSTRAP")) != nullptr ||
+      UniquePtr<char>(gpr_getenv("GRPC_XDS_BOOTSTRAP_CONFIG")) != nullptr) {
+    using_dns_ = true;
+    child_resolver_ = ResolverRegistry::CreateResolver(
+        absl::StrCat("dns:", name_to_resolve).c_str(), args.args,
+        args.pollset_set, work_serializer_, std::move(args.result_handler));
+    GPR_ASSERT(child_resolver_ != nullptr);
+    return;
+  }
+  // Create xds resolver.
+  child_resolver_ = ResolverRegistry::CreateResolver(
+      absl::StrCat("xds:", name_to_resolve).c_str(), args.args,
+      args.pollset_set, work_serializer_, std::move(args.result_handler));
+  GPR_ASSERT(child_resolver_ != nullptr);
+}
+
+void GoogleCloud2ProdResolver::StartLocked() {
+  if (using_dns_) {
+    child_resolver_->StartLocked();
+    return;
+  }
+  // Using xDS.  Start metadata server queries.
+  zone_query_ = MakeOrphanable<ZoneQuery>(Ref(), &pollent_);
+  ipv6_query_ = MakeOrphanable<IPv6Query>(Ref(), &pollent_);
+}
+
+void GoogleCloud2ProdResolver::RequestReresolutionLocked() {
+  if (child_resolver_ != nullptr) {
+    child_resolver_->RequestReresolutionLocked();
+  }
+}
+
+void GoogleCloud2ProdResolver::ResetBackoffLocked() {
+  if (child_resolver_ != nullptr) {
+    child_resolver_->ResetBackoffLocked();
+  }
+}
+
+void GoogleCloud2ProdResolver::ShutdownLocked() {
+  zone_query_.reset();
+  ipv6_query_.reset();
+  child_resolver_.reset();
+}
+
+void GoogleCloud2ProdResolver::ZoneQueryDone(std::string zone) {
+  zone_query_.reset();
+  zone_ = std::move(zone);
```

Comment:
since in case of error we will be filling in `zone_` with the empty string, does TD understand an empty string zone?